pennylane/devices/default_qubit.py

```diff
@@ -679,12 +729,24 @@ def _apply_unitary(self, state, mat, wires):
 
         Returns:
             array[complex]: output state
+
+        Note: This function does not support simultaneously broadcasted states and matrices yet.
         """
         # translate to wire labels used by device
         device_wires = self.map_wires(wires)
 
-        mat = self._cast(self._reshape(mat, [2] * len(device_wires) * 2), dtype=self.C_DTYPE)
-        axes = (np.arange(len(device_wires), 2 * len(device_wires)), device_wires)
+        dim = 2 ** len(device_wires)
+        mat_batch_size = self._get_batch_size(mat, (dim, dim), dim**2)
+        state_batch_size = self._get_batch_size(state, (2,) * self.num_wires, 2**self.num_wires)
+
+        shape = [2] * (len(device_wires) * 2)
+        state_axes = device_wires
+        if mat_batch_size:
+            shape.insert(0, mat_batch_size)
+        if state_batch_size:
+            state_axes = [ax + 1 for ax in state_axes]
+        mat = self._cast(self._reshape(mat, shape), dtype=self.C_DTYPE)
+        axes = (np.arange(-len(device_wires), 0), state_axes)
```

Comment:
If the matrix is broadcasted, it is reshaped to a shape that includes `mat_batch_size` as leading dimension.
if the state is broadcasted, the affected state axes need to be shifted by `1`.